archive/2025-08-05-resilience-by-design-security-in-the-age-of-offensive-ai-palo-alto-networks-blog.md

```diff
@@ -0,0 +1,257 @@
+
+Relying on a "scan-and-patch" cycle is like planning a week-long fortification project in the middle of an ambush.
+
+Our industry has poured immense resources into building an impenetrable fortress. We shifted left, armed developers with static application security testing (SAST) tools and hunted for every flaw. The _secure by design_ philosophy, for all its good intentions, is predicated on achieving a state of near-perfect invulnerability. The pursuit of perfection, of course, becomes a fool’s errand when an adversary can find and exploit an unknown flaw in the time it takes you to read this sentence. The wall will fail. The defining question is what will happen in the five minutes after that.
+
+## The Pivot to Engineering for Resilience
+
+Since we can’t guarantee perfect prevention, we need to shift our strategy from chasing invulnerability to engineering for security synergy. The new mandate is _resilience by design_.
+
+Visualize this approach using the [swiss cheese model](https://en.wikipedia.org/wiki/Swiss_cheese_model). Every defensive layer in our security stack, including code scanning, IAM policies, network segmentation and runtime protection, is a slice of cheese with unavoidable holes. An attacker succeeds only when holes in every layer align to create a direct path.
+
+A resilient architecture based on the Swiss cheese model relies on multiple, diverse defense layers. A vulnerability missed by a static application security testing scanner becomes useless under a least-privilege IAM role. A misconfigured IAM role receives neutralization from a runtime agent that detects and blocks anomalous behavior. Layer synergy prevents hole alignment.
+
+![Swiss cheese model for cloud-native security](https://www.paloaltonetworks.com/blog/wp-content/uploads/2025/08/word-image-342823-1.png)
+
+Figure 1: The swiss cheese model for cloud-native security
+
+A resilient architecture built according to this principle requires a fundamental change in how we build software and measure security success. It stands on four critical concepts.
+
+### Aggressive Containment
+
+Modern cloud-native systems must be architected as a series of segmented, isolated compartments to prevent a breach in one microservice or container from cascading.
+
+The most effective way to achieve isolation is by assigning each workload a strictly scoped, least-privileged IAM role. A compromised container then becomes a dead end for the attacker. Imagine a service whose IAM role grants only read access to a single S3 bucket. A breach of that service remains instantly contained. The attacker can’t access other databases, spin up new VMs or retrieve secrets. Lateral movement stops before it starts. I demonstrated this principle years ago in a Dark Reading article, detailing how a [compromised Lambda function becomes neutralized by a minimal IAM role](https://www.darkreading.com/cloud-security/securing-serverless-attacking-an-aws-account-via-a-lambda-function).
```

Comment:
A few years ago I found a vuln in Zapier's custom code execution environment. It was a lambda function, shared across users, so by hijacking it I was able to escalate privileges to an admin account. This point is very true. It's true beyond IAM to every architecture choice that limits the blast radius of an eventual compromise.